splash/html_element.py

```diff
@@ -0,0 +1,285 @@
+
+        if pad:
+            if isinstance(pad, float):
+                pad = (pad, pad, pad, pad)
+            region = (region[0] - pad[0], region[1] - pad[1], region[2] + pad[2], region[3] + pad[3])
+
+        return self.tab.png(width, region=region, scale_method=scale_method)
+
+    def jpeg(self, width=None, scale_method=None, quality=None, pad=None):
+        """ Return screenshot of the element in JPEG format """
+        self.assert_element_exists()
+
+        if not self.visible():
+            return None
+
+        bounds = self.bounds()
+        region = (bounds["left"], bounds["top"], bounds["right"], bounds["bottom"])
+
+        if pad:
+            if isinstance(pad, float):
+                pad = (pad, pad, pad, pad)
+            region = (region[0] - pad[0], region[1] - pad[1], region[2] + pad[2], region[3] + pad[3])
+
+        return self.tab.jpeg(width, region=region, scale_method=scale_method, quality=quality)
+
+    def visible(self):
+        """ Return flag indicating whether element is visible """
+        self.assert_element_exists()
+        return self.tab.evaljs(u"({visible_func})({element})".format(
+            visible_func=VISIBLE_JS_FUNC,
+            element=self.element_js
+        ))
+
+    def text(self):
+        """ Return text of the element """
+        return self.tab.evaljs(FETCH_TEXT_JS_FUNC % self.element_js)
+
+    def info(self):
+        """ Return information about the element """
+        return self.tab.evaljs(u"({element_info_func})({element}, {visible_func})".format(
+            element_info_func=ELEMENT_INFO_JS,
+            element=self.element_js,
+            visible_func=VISIBLE_JS_FUNC
+        ))
+
+    def field_value(self):
+        """ Return the value of the element if it is a field """
+        return self.tab.evaljs(u"({field_value_func})({element})".format(
+            field_value_func=FIELD_VALUE_JS,
+            element=self.element_js
+        ))
+
+    def form_values(self):
+        """ Return all values of the element if it is a form"""
+        self.assert_node_type('form')
+
+        return self.tab.evaljs(u"({form_values_func})({element}, {field_value_func})".format(
+            form_values_func=FORM_VALUES_JS,
+            field_value_func=FIELD_VALUE_JS,
+            element=self.element_js
+        ))
+
+    def fill(self, values, selector_type="names"):
+        """ Fill the values of the element """
+        if selector_type != "names":
+            raise NotImplemented('Only "names" selector type is supported')
+
+        return self.tab.evaljs(u"({fill_form_values_func})({element}, {values})".format(
+            fill_form_values_func=FILL_FORM_VALUES_JS,
+            element=self.element_js,
+            values=escape_js(values)
+        ))
+
+    def send_keys(self, text):
+        """ Send key events to the element separated by whitespaces """
+        self.mouse_click()
+        self.tab.send_keys(text)
```

Comment:
I remembered why I used `mouse_click` here. If try to use `focus` using `self.node_method('focus')` it doesn't work every time. Because when you call `node_method` it evaluates JS which is not immediate operation. On the contrary, calling `mouse_click` generates mouse click event using Qt.